code-gen/src/main/scala/scalapb/validate/NumericRulesPrinter.scala

```diff
@@ -0,0 +1,504 @@
+  private val JAVA_LONG_PKG = "java.lang.Long"
+  private val JAVA_FLOAT_PKG = "java.lang.Float"
+  private val JAVA_DOUBLE_PKG = "java.lang.Double"
+
+  private def gt[T](
+      fieldDescriptor: FieldDescriptor,
+      inputExpr: String,
+      defined: Boolean,
+      value: T,
+      typeParam: String
+  ): Option[String] =
+    if (defined)
+      Some(
+        s"""$COMPARATIVE_VALIDATION.greaterThan[$typeParam]("${fieldDescriptor.getName}", $inputExpr, $value, java.util.Comparator.naturalOrder)"""
+      )
+    else None
+
+  private def gte[T](
+      fieldDescriptor: FieldDescriptor,
+      inputExpr: String,
+      defined: Boolean,
+      value: T,
+      typeParam: String
+  ): Option[String] =
+    if (defined)
+      Some(
+        s"""$COMPARATIVE_VALIDATION.greaterThanOrEqual[$typeParam]("${fieldDescriptor.getName}", $inputExpr, $value, java.util.Comparator.naturalOrder)"""
+      )
+    else None
+
+  private def lt[T](
+      fieldDescriptor: FieldDescriptor,
+      inputExpr: String,
+      defined: Boolean,
+      value: T,
+      typeParam: String
+  ): Option[String] =
+    if (defined)
+      Some(
+        s"""$COMPARATIVE_VALIDATION.lessThan[$typeParam]("${fieldDescriptor.getName}", $inputExpr, $value, java.util.Comparator.naturalOrder)"""
+      )
+    else
+      None
+
+  private def lte[T](
+      fieldDescriptor: FieldDescriptor,
+      inputExpr: String,
+      defined: Boolean,
+      value: T,
+      typeParam: String
+  ): Option[String] =
+    if (defined)
+      Some(
+        s"""$COMPARATIVE_VALIDATION.lessThanOrEqual[$typeParam]("${fieldDescriptor.getName}", $inputExpr, $value, java.util.Comparator.naturalOrder)"""
+      )
+    else None
+
+  private def const[T](
+      fieldDescriptor: FieldDescriptor,
+      inputExpr: String,
+      defined: Boolean,
+      value: T,
+      typeParam: String
+  ): Option[String] =
+    if (defined)
+      Some(
+        s"""$CONSTANT_VALIDATION.constant[$typeParam]("${fieldDescriptor.getName}", $inputExpr, $value)"""
+      )
+    else None
+
+  private def toIterable(sq: Seq[Option[String]]): Iterable[String] = sq.flatten
+
+  def printFixed32Rules(
+      fieldDescriptor: FieldDescriptor,
+      inputExpr: String,
+      rules: Validate.Fixed32Rules
+  ): Iterable[String] =
+    toIterable(
+      Seq(
+        gt(fieldDescriptor, inputExpr, rules.hasGt, rules.getGt, JAVA_INT_PKG),
+        gte(
+          fieldDescriptor,
+          inputExpr,
+          rules.hasGte,
+          rules.getGte,
+          JAVA_INT_PKG
+        ),
+        lt(fieldDescriptor, inputExpr, rules.hasLt, rules.getLt, JAVA_INT_PKG),
+        lte(
+          fieldDescriptor,
+          inputExpr,
+          rules.hasLte,
+          rules.getLte,
+          JAVA_INT_PKG
+        ),
+        const(
+          fieldDescriptor,
+          inputExpr,
+          rules.hasConst,
+          rules.getConst,
+          JAVA_INT_PKG
+        )
+      )
+    )
+
+  def printFixed64Rules(
```

Comment:
I think all these functions can be deduped by using structural types:

```
type NumericRules[T] {
  def hasGte(): Boolean
  def getGte(): T
  // repeat for all other expected functions
}
```

then the same function can be re-used for all types without copy-paste